### PR TITLE
Cancel search on shard failure when partial results disallowed

### LINKED ISSUE #60278: Cancel queries on shard failures when partial results are not allowed
We have an option called `allow_partial_search_results`, that when set to `false` throws an error if all replicas of at least one shard fail. When the error happens, we continue the search until all shards respond and then throw the error at the end. That's a waste of resources to continue a search that we know is in error so we should try to cancel the request as soon as we detected that a shard is in error. That would make the handling of shard failures more reactive when the option is set.

### PR DESCRIPTION
**Pull request by dnhatn**
If the partial results parameter is false (which defaults to true), then we should cancel the search request when it hits a shard failure because the caller won't consume the partial results.

Closes #60278

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java [hunk 1]
```diff
@@ -393,6 +395,16 @@ private void onShardFailure(final int shardIndex, @Nullable SearchShardTarget sh
         logger.debug(() -> new ParameterizedMessage("{}: Failed to execute [{}] lastShard [{}]",
             shard != null ? shard : shardIt.shardId(), request, lastShard), e);
         if (lastShard) {
+            if (request.allowPartialSearchResults() == false) {
+                if (requestCancelled.compareAndSet(false, true)) {
+                    try {
+                        searchTransportService.cancelSearchTask(task, new CancellationException(
```
Comment (jimczi) [inline server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java:401]: The exception seems off, we should just set the error message or at least use a TaskCancelledException ?

**Review by jimczi — COMMENTED** [review]
I left some minor comments but the change makes sense to me.

File: server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java (hunk 1 above)
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java]: Good catch. Adjusted in https://github.com/elastic/elasticsearch/pull/63520/commits/f038c4388d3e3dd5d8c87df1eacfa92b09343c59

File: server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java [hunk 2]
```diff
@@ -423,4 +430,13 @@ private boolean assertNodePresent() {
             return true;
         }
     }
+
+    public void cancelSearchTask(SearchTask task, Exception exc) {
```
[the remark this answers is not in the document]
Comment (dnhatn) [reply server/src/main/java/org/elasticsearch/action/search/SearchTransportService.java]: ++

**Review by jimczi — APPROVED** [review]
LGTM

**Comment by dnhatn** [comment]
run elasticsearch-ci/2

**Pull request merged**